packages/go_router/lib/src/route.dart

```diff
@@ -320,6 +325,49 @@ class GoRoute extends RouteBase {
   late final RegExp _pathRE;
 }
 
+/// Base class for classes that act as shells for sub-routes, such
+/// as [ShellRoute] and [StatefulShellRoute].
+abstract class ShellRouteBase extends RouteBase {
+  /// Constructs a [ShellRouteBase].
+  const ShellRouteBase._({super.routes}) : super._();
+
+  /// Attempts to build the Widget representing this shell route.
+  ///
+  /// Returns null if this shell route does not build a Widget, but instead uses
+  /// a Page to represent itself (see [buildPage]).
+  Widget? buildWidget(BuildContext context, GoRouterState state,
+      ShellNavigatorBuilder navigatorBuilder);
+
+  /// Attempts to build the Page representing this shell route.
+  ///
+  /// Returns null if this shell route does not build a Page, but instead uses
+  /// a Widget to represent itself (see [buildWidget]).
+  Page<dynamic>? buildPage(BuildContext context, GoRouterState state,
+      ShellNavigatorBuilder navigatorBuilder);
+
+  /// Returns the key for the [Navigator] that is to be used for the specified
+  /// immediate sub-route of this shell route.
+  GlobalKey<NavigatorState> navigatorKeyForSubRoute(RouteBase subRoute);
+}
+
+/// Navigator builder for shell routes.
+abstract class ShellNavigatorBuilder {
+  /// The [GlobalKey] to be used by the [Navigator] built for the current route.
+  GlobalKey<NavigatorState> get navigatorKey;
```

Comment:
I think we should pass in subRoute instead of navigator. Not all shellRouteBase will build multiple navigator. Or just index to make thing more clear

Reply:
Yes, I guess we could just pass the subRoute - passing the navigatorKey was just a way of avoiding double-work. But it might be a better API to pass the subRoute.